File: python/pyfury/tests/benchmark.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import timeit
import pickle
import pytest
import pyfury as fury
from pyfury.tests.record import create_foo, foo_schema

iter_nums = 100000


@pytest.mark.skip(reason="take too long")
def test_encode():
    # print("schema", foo_schema())
    encoder = fury.create_row_encoder(foo_schema())
    foo = create_foo()
    row = encoder.to_row(foo)
    assert foo == encoder.from_row(row)

    t1 = timeit.timeit(lambda: encoder.to_row(foo), number=iter_nums)
    print(
        "encoder take {0} for {1} times, avg: {2}".format(t1, iter_nums, t1 / iter_nums)
    )
    t2 = timeit.timeit(lambda: pickle.dumps(foo), number=iter_nums)
    print(
        "pickle take {0} for {1} times, avg: {2}".format(t2, iter_nums, t2 / iter_nums)
    )


@pytest.mark.skip(reason="take too long")
def test_decode():
    # print(foo_schema()
    encoder = fury.create_row_encoder(foo_schema())
    foo = create_foo()

    row = encoder.to_row(foo)
    assert foo == encoder.from_row(row)
    t1 = timeit.timeit(lambda: encoder.from_row(row), number=iter_nums)
    print(
        "encoder take {0} for {1} times, avg: {2}, size {3}".format(
            t1, iter_nums, t1 / iter_nums, row.size_bytes()
        )
    )
    pickled_data = pickle.dumps(foo)
    t2 = timeit.timeit(lambda: pickle.loads(pickled_data), number=iter_nums)
    print(
        "pickle take {0} for {1} times, avg: {2}, size {3}".format(
            t2, iter_nums, t2 / iter_nums, len(pickled_data)
        )
    )


if __name__ == "__main__":
    test_encode()
    test_decode()

File: java/benchmark/src/main/java/org/apache/fury/benchmark/Benchmark.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fury.benchmark;

import org.openjdk.jmh.Main;

public class Benchmark {
  // run from cmd:
  // cd .. && mvn -T10 install -DskipTests -Dcheckstyle.skip -Dlicense.skip -Dmaven.javadoc.skip
  // mvn exec:java -Dexec.args="-f 0 -wi 1 -i 1 -t 1 -w 1s -r 1s -rf csv"
  // add `-XX:+PrintCompilation -XX:+UnlockDiagnosticVMOptions -XX:+PrintInlining` to inspect
  // the compile process.
  public static void main(String[] args) throws Exception {
    if (args.length == 0) {
      String commandLine =
          "org.apache.fury.*_deserialize$ -f 3 -wi 5 -i 3 -t 1 -w 3s -r 3s -rf csv";
      System.out.println(commandLine);
      args = commandLine.split(" ");
    }
    Main.main(args);
  }
}
